### PR TITLE
RFC: generate PropTypes, handledProps

### PR DESCRIPTION
**Pull request by layershifter**
# RFC

Hey there!

I've made a dirty POC to check that it's possible. My proposal contains some usefull things.

### `propTypes`
Addresses #7. I almost sure that we should generate `propTypes` from interfaces, this PR is an initial POC for this. We also can to add there the environment checks.

### `handledProps`

As we use AST transforms, we can easily add `handledProps`, too 👍 

----

This PR heavily relies on [ttypescript](https://github.com/cevek/ttypescript), please let me know what you think about this.

### CODE REVIEW

**Comment by kuzhelov** [comment]
**TL;DR** absolute support for generating propTypes if it would be possible to do robustly, more complicated story for `handledProps`.

----------
Personally totally support idea of generating prop types from TS types - in case if it would be possible to do in robust way.

Speaking of `handledProps` - the story is a bit more complicated here, as at some points it is about deliberate decision of the developer whether property defined in PropTypes should be introduced to the set of handled ones - and, thus, won't be passed down to component's child tree. However, as `handledProps` in **_vast majority of cases_** replicate the set of PropTypes, we could think about introducing property with inverted semantics instead of `handledProps` - something like `unhandledProps`. In that case `unhandledProps` array will be an empty one for most of components, and only in case if property needs to be passed to component's child tree - only in this case the property should be added to this array.

Please, let me know about your thoughts on that :)

**Comment by levithomason** [comment]
Vote: Yes, generate a `handledProps` array from our source.

***

I think the inverted case will be too large.  We cannot be sure of every possible prop name a user might want to pass down.  They may be integrating with some special 3rd party work and need to pass obscure names down.  I would propose we keep the list of "consumed" or "handled" props.

As for generating, we've generated the `handledProps` in Semantic UI React for a few years now and it seems to have served us very well.  I don't believe we've had any issues in this area regarding the generation (babel plugin @layershifter wrote for us) or the pattern of passing all unhandled props down.

**Comment by levithomason** [comment]
>This PR heavily relies on ttypescript, please let me know what you think about this.

It seems like an obscure package and I'd be wary of relying too much of things that may vanish at any point.  That said, so long as we had robust tests around this as in Semantic UI React, then I think it would be OK.

Thoughts?

**Comment by layershifter** [comment]
> It seems like an obscure package and I'd be wary of relying too much of things that may vanish at any point. 

Yep, it is. I am also not happy with this solution. As an option, we can move its functionality there.

We can also try the another solution, too. Babel 7 supports Typescript, we can restore it back and use [babel-plugin-typescript-to-proptypes](https://github.com/milesj/babel-plugin-typescript-to-proptypes) with other plugins that we use in SUIR.

**Comment by kuzhelov** [comment]
@levithomason 

> I think the inverted case will be too large. We cannot be sure of every possible prop name a user might want to pass down. They may be integrating with some special 3rd party work and need to pass obscure names down. I would propose we keep the list of "consumed" or "handled" props.

actually, I am a bit unsure why it would be a problem. Lets consider concrete case, just to be sure that we are talking about the same logic proposed. Suppose that we have an Icon component with PropTypes `circular` and `bordered`. In that case, with the following way of declaring it (with no '(un)handledProps') both `circular` and `bordered` won't be passed down to children. 

```jsx
class Icon ... {
    propTypes = {
        circular: PropTypes.bool,
        bordered: PropTypes.bool
    }
...
}
```

At the same time, all other properties (that doesn't belong to set of declared prop types) **_will be passed_**. As you might notice, this is a behavior that we currently have for vast majority of components - but, in contrast to what we do have now, we don't need to repeat props in `handledProps` array.

Now, lets suppose that we would need to pass some of propTypes down to component's children. In that case corresponding property should be added to `unhandledProps` array:

```jsx
class Icon ... {
    propTypes = {
        circular: PropTypes.bool,
        bordered: PropTypes.bool
    }

    unhandledProps = [ 'circular' ]
}
```
with these changes `circular` property will be passed down to Icon's child. 

------

As you might see, this approach will, essentially, preserve the semantics that we currently have - while, at the same time, it will require much less typing (because currently for almost all of the components we have `handledProps` as duplication of those declared in `propTypes`).

**Comment by layershifter** [comment]
@kuzhelov we have a multiple places where we need to pass down some of `handledProps` in SUIR. Why don't pass them manually? It looks like more obvious solution.

```jsx
class Icon ... {
  propTypes = {
    circular: PropTypes.bool,
    bordered: PropTypes.bool
  }

  renderElement() {
    const { bordered, circular } = this.props

    return (
      <ElementType circular={circular} {...rest}>
        <Child bordered={bordered} />
      </ElementType>
    )
  }
}
```

It also breaks the `ElementType` pattern, because an `as` prop value can HTML tag and we will receive a React warning.

-----

BTW, we discussed the similar idea with Levi about props sources, https://github.com/layershifter/babel-plugin-transform-react-handled-props/issues/6#issuecomment-274964318. It's still unnecessary feature 😺

**Comment by layershifter** [comment]
The reason of usage `ttypescript` is there, https://github.com/ivogabe/gulp-typescript/issues/502. With `gulp-typescript` we cannot modify options passed to passed `Program.emit()`.

In the same time, `awesome-typescript-loader` supports it, [getCustomTransformers()](https://github.com/s-panferov/awesome-typescript-loader#getcustomtransformers-string--program-tsprogram--tscustomtransformers--undefined-defaultundefined).

The reasonable resolution that I see it's contribute to `gulp-typescript` and get working `getCustomTransformers` option there. Will work on it 🚀

**Comment by layershifter** [comment]
Tracking https://github.com/ivogabe/gulp-typescript/pull/589.

**Comment by layershifter** [comment]
My PR was merged (ivogabe/gulp-typescript#589) :v: But, there are no release now, waiting for it.

**Comment by layershifter** [comment]
We decided to switch to Babel, closing for housekeeping.
